### ISSUE TITLE
Pools with `minCapacity` set interact poorly with `shutdown.afterIdleSeconds`/`idleTimeoutSecs`.

### ISSUE THREAD

**Issue by tomprince** [taskcluster/taskcluster#3072]
**Describe the bug**

A worker with `shutdown.afterIdleSeconds`/`idleTimeoutSecs` will shutdown if it does not receive work from the queue in the specified amount of time. This interacts poorly with pools that have a `minCapacity` set, as worker manager will immediately spin up a new worker with cold caches. It would be much better if enough workers to satisfy `minCapacity` stayed alive, rather then endlessly cycling.

This can have a huge impact on gecko action tasks, where the cost of a fresh clone ends up taking about half the runtime of a task.

(cc: @jmaher)

**Comment by jmaher** [comment]
I intentionally lowered the shutdown.afterIdleSeconds to be test machines, not machines with source checkout.  With that said I imagine we can fine tune this for all workers.

**Comment by djmitche** [comment]
The idle decision is made by the worker, whereas minCapacity requires a knowledge of the total running capacity and is made in the worker-manager service.  So this requires some kind of coordination between the two, and some way of deciding, given a set of N workers with N > minCapacity, which of those N should not timeout.  That's not insolvable, but not simple either.

It might be better to focus on the goal rather than trying to reconcile these two configuration parameters.  A situation where, say, three workers run forever but the fourth worker terminates immediately on idle seems rather precipitous.  I think the goal here is rapid task start-up and warm caches for a spiky load, with limited idleness.  Maybe there's another model that can do a better job of auto-tuning itself to the load?

**Comment by jmaher** [comment]
the goal is to reduce our non task overhead as much as possible.  We find that most workers types have 50%+ of the billing as idle time falling into a few buckets:
1) hung workers
2) over provisioned workers
3) bootstrap time
4) idle between tasks
5) idle and shutdown after all tasks

Ideally we would increase the workers when the queue is filling up waiting on builds.  Can we set a queue size that we accept.  Say we have 100 jobs for a workertype in the queue, then we spin up 5 workers.  I don't know why need to have workers always ready- there might be some cases like decision task where we need to have workers available faster as delays there block everything else.  Builds might have different thresholds than testers, etc.

**Comment by djmitche** [comment]
Yep, there are a mix of priorities here -- having workers ready for tasks (E2E and time-to-first-task), keeping caches warm, and reducing overhead all seem relevant here.  Optimizing any one or two of them is pretty easy -- for example for testers if we just ran one permanently, that'd optimize overhead and caches but obviously be pretty bad for time-to-first-task.

I agree that there are different conditions for different pools -- decision tasks want to prefer a short E2E, builders need warm caches, testers need to reduce overhead.  And those pools have different usage patterns, too -- testers get 100's of tasks marked pending all at once, when the single build those tasks depend on completes, and those tasks tend to be fairly short-duration, whereas docker-image builds are rare but tend to be fairly long duration (and also block lots of other tasks so are E2E-sensitive).

Right now, the main knobs are idle timeout and `minCapacity`.  We've talked vaguely about knobs that might control the rate of spawning of new workers, and approaches to model and predict the workload based on information from the queue.  That's a difficult design problem, but would be pretty high-impact if we could get it right.

I say all this to highlight that I think there's a higher-level problem, or set of problems, to solve here, and that we are interested in solving.  The engineering effort required to get idle timeout and `minCapacity` to play nicely together is high and probably does not address that larger problem very well.

**Comment by tomprince** [comment]
> the goal is to reduce our non task overhead as much as possible.

Time taken in vcs (and fetching other things that could be cached) is also overhead, but counted as part of tasks. I guess idle time between jobs probably dominates for most workers, so it is probably always a win in practice; but decreasing non-task overhead can lead to longer task times.

>  don't know why need to have workers always ready- there might be some cases like decision task where we need to have workers available faster as delays there block everything else.

Yeah, decision workers at least want reasonably fast turn-around times; particularly as these are also used for actions.

> I say all this to highlight that I think there's a higher-level problem, or set of problems, to solve here, and that we are interested in solving. The engineering effort required to get idle timeout and `minCapacity` to play nicely together is high and probably does not address that larger problem very well.

Yeah, I definitely don't know what the right solution is, and it probably needs to take other related aspects of provisioning. I do think it is worth calling out explicitly, even if the eventual solution ends up changing things so that one or the other or both these knobs don't exist in their current form. I can also imagine that (in the context of a large solution), solving this might be an (comparatively) easy detour part way along the implementation of the full solution.

**Comment by jmaher** [comment]
Should we setup a meeting to understand this problem more and start
thinking about solutions?  I imagine the work that we end up doing will be
really awesome yet much different than what one of us would propose right
now.

**Comment by Eijebong** [comment]
This was fixed by #8327, minCapacity now works as intended. This was deployed on fxci on march 5th and had a very visible impact on decision task timing (the weekend spikes due to low worker presence are completely gone)
